src/_pytest/runner.py

```diff
@@ -190,35 +191,45 @@ def check_interactive_exception(call, report):
 def call_runtest_hook(item, when, **kwds):
     hookname = "pytest_runtest_" + when
     ihook = getattr(item.ihook, hookname)
-    return CallInfo(
+    return CallInfo.from_call(
         lambda: ihook(item=item, **kwds),
         when=when,
-        treat_keyboard_interrupt_as_exception=item.config.getvalue("usepdb"),
+        reraise=KeyboardInterrupt if item.config.getvalue("usepdb") else (),
     )
 
 
+@attr.s
 class CallInfo(object):
     """ Result/Exception info a function invocation. """
 
-    #: None or ExceptionInfo object.
-    excinfo = None
-
-    def __init__(self, func, when, treat_keyboard_interrupt_as_exception=False):
+    _result = attr.ib()
+    # type: Optional[ExceptionInfo]
+    excinfo = attr.ib()
+    start = attr.ib()
+    stop = attr.ib()
+    when = attr.ib()
+
+    @property
+    def result(self):
+        if self.excinfo is not None:
+            raise AttributeError("{!r} has no valid result".format(self))
+        return self._result
+
+    @classmethod
+    def from_call(cls, func, when, reraise=None):
         #: context of invocation: one of "setup", "call",
         #: "teardown", "memocollect"
-        self.when = when
-        self.start = time()
+        start = time()
+        excinfo = None
         try:
-            self.result = func()
-        except KeyboardInterrupt:
-            if treat_keyboard_interrupt_as_exception:
-                self.excinfo = ExceptionInfo()
-            else:
-                self.stop = time()
-                raise
+            result = func()
         except:  # noqa
-            self.excinfo = ExceptionInfo()
-        self.stop = time()
+            excinfo = ExceptionInfo()
+            if reraise is not None and excinfo.errisinstance(reraise):
+                raise
+            result = None
+        stop = time()
+        return cls(start=start, stop=stop, when=when, result=result, excinfo=excinfo)
```

Comment:
Now it re-raises `KeyboardInterrupt`, but previously it was swallowed.

```diff
diff --git i/src/_pytest/runner.py w/src/_pytest/runner.py
index 44905fd7..c20d077a 100644
--- i/src/_pytest/runner.py
+++ w/src/_pytest/runner.py
@@ -194,7 +194,7 @@ def call_runtest_hook(item, when, **kwds):
     return CallInfo.from_call(
         lambda: ihook(item=item, **kwds),
         when=when,
-        reraise=KeyboardInterrupt if item.config.getvalue("usepdb") else (),
+        reraise=KeyboardInterrupt if not item.config.getvalue("usepdb") else (),
     )
```

But reraises should be an iterable always maybe?

Reply:
argh